File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/conv_winograd_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/conv_winograd_emit.py -- Python reference emitter for the
# Winograd convolution parity harness.
#
# Selects one of N sampled spec configs by argv[1], builds a WinogradConvSpec,
# builds all three transform kernels (data / filter / output) via the
# build_winograd_* builders, and prints lower_kernel_to_llvm(arch=<cfg arch>)
# for the selected kernel to stdout so it can be byte-compared with the C
# emitter conv_winograd_emit.c.
#
# argv[2] selects the sub-kernel within the config:
#   "data"   -- data transform kernel   (default)
#   "filter" -- filter transform kernel
#   "output" -- output transform kernel
#
# NOTE: The C counterpart (conv_winograd_emit.c) is a placeholder stub until
# the C++ engine mirror is implemented. Until then, run_diff.py will report
# this family as UNSUPPORTED on the C side, which is the gate-passing status
# for new families not yet mirrored in C++.

from rocke.instances.common.conv_winograd import (
    WinogradConvSpec,
    WinogradProblem,
    build_winograd_data_transform,
    build_winograd_filter_transform,
    build_winograd_output_transform,
)
from _emit_common import run_emit
import sys


def _spec(idx: int):
    """Return (spec, arch, sub_kernel) for config index `idx`.

    Sub-kernel is one of "data", "filter", "output"; it is baked into the
    config index so run_diff.py can treat each transform as a separate config.
    Each logical problem contributes three consecutive indices:
        idx 0,1,2  -- F(4,3) N8 H56 W56 C64 K64  data/filter/output
        idx 3,4,5  -- F(2,3) N8 H56 W56 C64 K64  data/filter/output
        idx 6,7,8  -- F(4,3) N4 H28 W28 C128 K128  data/filter/output
        idx 9,10,11-- F(4,3) N1 H7 W7 C512 K512  data/filter/output
    """
    _sub = ["data", "filter", "output"]

    if 0 <= idx <= 2:
        p = WinogradProblem(N=8, Hi=56, Wi=56, C=64, K=64, pH=1, pW=1)
        spec = WinogradConvSpec(
            problem=p, out_tile=4, block_c=32, block_k=32, block_nhw=4
        )
        return spec, "gfx950", _sub[idx % 3]

    if 3 <= idx <= 5:
        p = WinogradProblem(N=8, Hi=56, Wi=56, C=64, K=64, pH=1, pW=1)
        spec = WinogradConvSpec(
            problem=p, out_tile=2, block_c=32, block_k=32, block_nhw=4
        )
        return spec, "gfx950", _sub[idx % 3]

    if 6 <= idx <= 8:
        p = WinogradProblem(N=4, Hi=28, Wi=28, C=128, K=128, pH=1, pW=1)
        spec = WinogradConvSpec(
            problem=p, out_tile=4, block_c=32, block_k=32, block_nhw=4
        )
        return spec, "gfx950", _sub[idx % 3]

    if 9 <= idx <= 11:
        p = WinogradProblem(N=1, Hi=7, Wi=7, C=512, K=512, pH=1, pW=1)
        spec = WinogradConvSpec(
            problem=p, out_tile=4, block_c=32, block_k=32, block_nhw=1
        )
        return spec, "gfx942", _sub[idx % 3]

    return None


def _build(idx: int):
    """Return (kernel, arch) for the given config index."""
    result = _spec(idx)
    if result is None:
        return None
    spec, arch, sub = result
    builders = {
        "data": build_winograd_data_transform,
        "filter": build_winograd_filter_transform,
        "output": build_winograd_output_transform,
    }
    kernel = builders[sub](spec, arch=arch)
    return kernel, arch


def _spec_only(idx: int):
    result = _spec(idx)
    if result is None:
        return None
    spec, arch, sub = result
    return spec, arch


def _build_kernel(spec, arch):
    # run_emit calls build_fn(spec, arch=arch); we ignore spec here since
    # _build() already encodes the sub-kernel selection by idx.
    raise NotImplementedError("use _build() directly")


# run_emit expects spec_fn(idx) -> (spec, arch) and build_fn(spec, arch=arch).
# For Winograd we bypass this by providing a wrapped spec_fn that returns the
# fully-built kernel directly.
if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.stderr.write("usage: conv_winograd_emit.py <config_index> [ll|ir|verify]\n")
        raise SystemExit(2)

    idx = int(sys.argv[1])
    mode = sys.argv[2] if len(sys.argv) > 2 else "ll"

    result = _build(idx)
    if result is None:
        # Signal "unknown config" — run_diff stops enumeration here.
        sys.stderr.write(f"unknown config {idx}\n")
        raise SystemExit(1)

    kernel, arch = result

    try:
        from rocke.core.lower_llvm import _lower_kernel_to_llvm_python as _native_lower
    except ImportError:
        from rocke import lower_kernel_to_llvm as _native_lower

    from rocke.core.ir_serialize import serialize
    from rocke.core.verify import verify

    if mode == "ll":
        sys.stdout.write(_native_lower(kernel, arch=arch))
    elif mode == "ir":
        sys.stdout.write(serialize(kernel))
    elif mode == "verify":
        sys.stdout.write(verify(kernel, arch=arch))
    else:
        sys.stderr.write(f"unknown mode {mode}\n")
        raise SystemExit(2)

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/conv_winograd_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/conv_winograd_emit.c -- C-side emitter for the Winograd
 * convolution parity harness.
 *
 * Selects one of N sampled spec configs by argv[1], builds the appropriate
 * Winograd transform kernel via the C++ engine, and lowers it to LLVM IR so
 * run_diff.py can byte-compare it with the Python reference
 * conv_winograd_emit.py.
 *
 * Config index layout — mirrors conv_winograd_emit.py exactly:
 *   idx 0,1,2  — F(4,3) N8 H56 W56 C64 K64  gfx950  data/filter/output
 *   idx 3,4,5  — F(2,3) N8 H56 W56 C64 K64  gfx950  data/filter/output
 *   idx 6,7,8  — F(4,3) N4 H28 W28 C128 K128 gfx950  data/filter/output
 *   idx 9,10,11— F(4,3) N1 H7 W7 C512 K512  gfx942  data/filter/output
 *
 * Sub-kernel: idx % 3 → 0=data, 1=filter, 2=output.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_conv_winograd.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Sub-kernel selector */
#define SUB_DATA 0
#define SUB_FILTER 1
#define SUB_OUTPUT 2

/* Fill spec and arch for config index idx.
 * Returns 0 on success, -1 if idx is out of range. */
static int make_cfg(int idx, rocke_winograd_conv_spec_t* spec, const char** arch, int* sub)
{
    *spec = rocke_winograd_conv_spec_default();
    *sub = idx % 3;

    switch(idx / 3)
    {
    case 0: /* F(4,3) N8 H56 W56 C64 K64 gfx950 */
        spec->problem = rocke_winograd_problem_default(8, 56, 56, 64, 64);
        spec->out_tile = 4;
        spec->block_c = 32;
        spec->block_k = 32;
        spec->block_nhw = 4;
        *arch = "gfx950";
        return 0;

    case 1: /* F(2,3) N8 H56 W56 C64 K64 gfx950 */
        spec->problem = rocke_winograd_problem_default(8, 56, 56, 64, 64);
        spec->out_tile = 2;
        spec->block_c = 32;
        spec->block_k = 32;
        spec->block_nhw = 4;
        *arch = "gfx950";
        return 0;

    case 2: /* F(4,3) N4 H28 W28 C128 K128 gfx950 */
        spec->problem = rocke_winograd_problem_default(4, 28, 28, 128, 128);
        spec->out_tile = 4;
        spec->block_c = 32;
        spec->block_k = 32;
        spec->block_nhw = 4;
        *arch = "gfx950";
        return 0;

    case 3: /* F(4,3) N1 H7 W7 C512 K512 gfx942 */
        spec->problem = rocke_winograd_problem_default(1, 7, 7, 512, 512);
        spec->out_tile = 4;
        spec->block_c = 32;
        spec->block_k = 32;
        spec->block_nhw = 1;
        *arch = "gfx942";
        return 0;

    default:
        return -1;
    }
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: conv_winograd_emit <config_index> [ll|ir|verify]\n");
        return 2;
    }

    int idx = atoi(argv[1]);
    const char* mode = (argc >= 3) ? argv[2] : "ll";

    if(strcmp(mode, "ll") != 0 && strcmp(mode, "ir") != 0 && strcmp(mode, "verify") != 0)
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        return 2;
    }

    rocke_winograd_conv_spec_t spec;
    const char* arch = NULL;
    int sub = 0;

    if(make_cfg(idx, &spec, &arch, &sub) != 0)
    {
        /* Signal "unknown config" — run_diff.py stops enumeration. */
        fprintf(stderr, "unknown config %d\n", idx);
        return 1;
    }

    rocke_ir_builder_t b;
    rocke_kernel_def_t* kernel = NULL;

    switch(sub)
    {
    case SUB_DATA:
        kernel = rocke_build_winograd_data_transform_new(&b, &spec, arch);
        break;
    case SUB_FILTER:
        kernel = rocke_build_winograd_filter_transform_new(&b, &spec, arch);
        break;
    case SUB_OUTPUT:
        kernel = rocke_build_winograd_output_transform_new(&b, &spec, arch);
        break;
    }

    if(kernel == NULL)
    {
        fprintf(stderr, "build failed for config %d: %s\n", idx, rocke_ir_builder_error(&b));
        rocke_ir_builder_free(&b);
        return 1;
    }

    int rc = 0;

    if(strcmp(mode, "ll") == 0)
    {
        char* ll_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, arch, &ll_text);
        if(st != ROCKE_OK || ll_text == NULL)
        {
            fprintf(stderr, "lower_kernel_to_llvm failed (status %d)\n", (int)st);
            rc = 1;
        }
        else
        {
            fputs(ll_text, stdout);
            free(ll_text);
        }
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* ir_text = NULL;
        rocke_status_t st2 = rocke_ir_serialize(kernel, &ir_text);
        if(st2 != ROCKE_OK || ir_text == NULL)
        {
            fprintf(stderr, "ir_serialize failed (status %d)\n", (int)st2);
            rc = 1;
        }
        else
        {
            fputs(ir_text, stdout);
            free(ir_text);
        }
    }
    else /* verify */
    {
        rocke_diag_t* diags = NULL;
        size_t n_diags = 0;
        rocke_verify(kernel, &diags, &n_diags);
        for(size_t i = 0; i < n_diags; ++i)
            fprintf(stdout, "%s\n", diags[i].message);
        rocke_diags_free(diags, n_diags);
    }

    rocke_ir_builder_free(&b);
    return rc;
}
